packages/driver/src/cy/snapshots.ts

```diff
@@ -10,6 +10,94 @@ export const HIGHLIGHT_ATTR = 'data-cypress-el'
+  // node.parentNode.host.shadowRoot would work. But this is considered an instance of an Object and not
+  // a ShadowRoot, so likely checking the string serialization might be a safer choice here.
+  const isNodeShadowRoot = (n: any) => n?.toString() === '[object ShadowRoot]'
+
+  let parent = node && node.parentNode
+
+  while (parent) {
+    if (isNodeShadowRoot(parent)) {
+      return parent as ShadowRoot
+    }
+
+    parent = parent.parentNode
+  }
+
+  return null
+}
+
+/**
+ *
+ * @param elem - either an HTML Element or an element that lives within the ShadowDom or the Root Document
+ * @returns SelectorNode if the selector can be discovered. For regular elements, this should only be one object deep, but for ShadowDom
+ * elements, the SelectorNode tree could be N levels deep until the root is discovered
+ */
+function constructElementSelectorTree (elem: Element): SelectorNode | undefined {
+  try {
+    const ownerDoc = elem.ownerDocument
+    const elWindow = ownerDoc.defaultView
+
+    if (elWindow === null) {
+      return undefined
+    }
+
+    // finder tries to find the shortest unique selector to an element,
+    // but since we are more concerned with speed, we set the threshold to 1 and maxNumberOfTries to 0
+    // @ts-expect-error because 'root' can be either Document or Element but is defined as Element
+    // @see https://github.com/antonmedv/finder/issues/75
+    const selector = finder(elem, { root: ownerDoc, threshold: 1, maxNumberOfTries: 0 })
+
+    if (!selector) {
+      throw new Error('Selector not defined! Fall through to shadowDom lookup')
+    }
+
+    const frameId = elWindow['__cypressProtocolMetadata']?.frameId
+
+    return { selector, frameId, ownerDoc: elem.ownerDocument, host: undefined }
+  } catch {
+    // the element may not always be found since it's possible for the element to be removed from the DOM
+    // Or maybe its in the shadow DOM.
+    // If it is a shadow DOM element, return the ShadowRoot as well to relate the node to the root document
+    try {
+      const shadowRoot = returnShadowRootIfShadowDomNode(elem)
+
+      // If we have a shadow DOM element, get the frameId and unique selector of the ShadowRoot
+      // see https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot
+      if (shadowRoot) {
+        if (shadowRoot.mode !== 'open') {
+          // cannot see into closed shadow DOM. Cannot reconstruct for Test Replay
+          // this code should NOT be reachable
+          return undefined
+        }
+
+        // Look up the details of the shadowRoot to see which element the ShadowRoot is bound to, i.e. the host.
+        const hostDetails = constructElementSelectorTree(shadowRoot.host)
+
+        // look up our element inside the context of the ShadowRoot
+        const selectorFromShadowWorld = finder(elem, { root: shadowRoot as unknown as Element, threshold: 1, maxNumberOfTries: 0 })
```

Comment:
Since we are repeating the `finder` options here, we might consider creating a new function for this logic so we don't accidentally change one but not the other.

Reply:
addressed in 9132ea1225